mlir/test/Mitigation/ZneFoldingAllTest.mlir

```diff
@@ -0,0 +1,87 @@
+    // CHECK:   [[%q2:%.+]] = quantum.namedobs [[q01_out2]]#0[ PauliY] : !quantum.obs
+    // CHECK:   [[results:%.+]] = quantum.expval [[q1]] : f64
+    // CHECK:   [[tensorRes:%.+]] = tensor.from_elements [[result]] : tensor<f64>
+    // CHECK:   [[%q2:%.+]] = quantum.insert %0[ 0], [[q01_out2]]#0 : !quantum.reg, !quantum.bit
+    // CHECK:   [[%q3:%.+]] = quantum.insert %7[ 1], [[q01_out2]]#1 : !quantum.reg, !quantum.bit
+    // CHECK:   quantum.dealloc [[q2]] : !quantum.reg
+    // CHECK:   quantum.device_release
+    // CHECK:   return [[tensorRes]]
+
+// CHECK:    func.func private @simpleCircuit.quantumAlloc(%arg0: i64) -> !quantum.reg {
+    // CHECK:    [[allocQreg:%.+]] = quantum.alloc(%arg0) : !quantum.reg
+    // CHECK:    return [[allocQreg]] : !quantum.reg
+
+//CHECK-LABEL: func.func @circuit
+func.func @circuit() -> tensor<f64> attributes {qnode} {
+    quantum.device ["rtd_lightning.so", "LightningQubit", "{shots: 0}"]
+    %0 = quantum.alloc( 2) : !quantum.reg
+    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
+    %out_qubits = quantum.custom "Hadamard"() %1 : !quantum.bit
+    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
+    %out_qubits_0:2 = quantum.custom "CNOT"() %out_qubits, %2 : !quantum.bit, !quantum.bit
+    %3 = quantum.namedobs %out_qubits_0#0[ PauliY] : !quantum.obs
+    %4 = quantum.expval %3 : f64
+    %from_elements = tensor.from_elements %4 : tensor<f64>
+    %5 = quantum.insert %0[ 0], %out_qubits_0#0 : !quantum.reg, !quantum.bit
+    %6 = quantum.insert %5[ 1], %out_qubits_0#1 : !quantum.reg, !quantum.bit
+    quantum.dealloc %6 : !quantum.reg
+    quantum.device_release
+    return %from_elements : tensor<f64>
+}
+
+
+//CHECK-LABEL: func.func @mitigated_circuit()
+    //CHECK:    [[c0:%.+]] = index.constant 0
+    //CHECK:    [[c1:%.+]] = index.constant 1
+    //CHECK:    [[c3:%.+]] = index.constant 3
+    //CHECK:    [[dense3:%.+]] = arith.constant dense<[1, 2, 3]>
+    //CHECK:    [[emptyRes:%.+]] = tensor.empty() : tensor<3xf64>
+    //CHECK:    [[results:%.+]] = scf.for [[idx:%.+]] = [[c0]] to [[c3]] step [[c1]] iter_args(%arg1 = [[emptyRes]]) -> (tensor<3xf64>) {
+        //CHECK:    [[scaleFactor:%.+]] = tensor.extract [[dense3]][[[idx]]] : tensor<3xindex>
+        //CHECK:    [[intermediateRes:%.+]] = func.call @circuit.folded([[scaleFactor]]) : (index) -> tensor<f64>
+        //CHECK:    [[tensorRes:%.+]] = tensor.from_elements [[intermediateRes]] : tensor<1xf64>
+        //CHECK:    [[resultsFor:%.+]] = scf.for %arg2 = [[c0]] to [[c1]] step [[c1]] iter_args(%arg3 = %arg1) -> (tensor<3xf64>) {
+            //CHECK:    [[extracted:%.+]] = tensor.extract [[tensorRes]][%arg3] : tensor<1xf64>
+            //CHECK:    [[insertedRes:%.+]] = tensor.insert [[extracted]] into %arg3[%arg1] : tensor<5xf64>
+            //CHECK:    scf.yield [[insertedRes]]
+        //CHECK:    scf.yield [[resultsFor]]
+    //CHECK:    return [[results]]
```

Comment:
Nit: Personally I think the following format is slightly more readable (indentation is definitely useful):
```suggestion
    //CHECK:    [[c0:%.+]] = index.constant 0
    //CHECK:    [[c1:%.+]] = index.constant 1
    //CHECK:    [[c3:%.+]] = index.constant 3
    //CHECK:    [[dense3:%.+]] = arith.constant dense<[1, 2, 3]>
    //CHECK:    [[emptyRes:%.+]] = tensor.empty() : tensor<3xf64>
    //CHECK:    [[results:%.+]] = scf.for [[idx:%.+]] = [[c0]] to [[c3]] step [[c1]] iter_args(%arg1 = [[emptyRes]]) -> (tensor<3xf64>) {
    //CHECK:        [[scaleFactor:%.+]] = tensor.extract [[dense3]][[[idx]]] : tensor<3xindex>
    //CHECK:        [[intermediateRes:%.+]] = func.call @circuit.folded([[scaleFactor]]) : (index) -> tensor<f64>
    //CHECK:        [[tensorRes:%.+]] = tensor.from_elements [[intermediateRes]] : tensor<1xf64>
    //CHECK:        [[resultsFor:%.+]] = scf.for %arg2 = [[c0]] to [[c1]] step [[c1]] iter_args(%arg3 = %arg1) -> (tensor<3xf64>) {
    //CHECK:            [[extracted:%.+]] = tensor.extract [[tensorRes]][%arg3] : tensor<1xf64>
    //CHECK:            [[insertedRes:%.+]] = tensor.insert [[extracted]] into %arg3[%arg1] : tensor<5xf64>
    //CHECK:            scf.yield [[insertedRes]]
    //CHECK:        scf.yield [[resultsFor]]
    //CHECK:    return [[results]]
```